### PR TITLE
Eng 1359 fix a bug where tables with integer

### PR DESCRIPTION
**Pull request by kenxu95**
## Describe your changes and why you are making these changes
We cannot support non-string column names for dataframes due to json-serialization (and also parquet if we chose to migrate), so if we detect such cases during serialization, we throw an error.

When you try to do something like this, you'll get:
image

In order to implement this nicely, we introduce the concept of an `ExecFailureException`, which can be translated to a failed `ExecutionState` at the top level. This allows us to not have to pass an exec_state around all the time to catch user errors. In this case, we had to throw an exception within the serialization helpers. It would have been somewhat to have to pass an exec_state all the way down.

In order to keep the two types of ExecutionState paths separate, we support the following pattern now, in addition to our `@user_fn_redirected` pattern:
```
raise ExecFailureException(...)
...
except ExecFailureException as e:
     write(ExecutionState.from_exception(e, user_logs=...))
```

@vsreekanti to inspect the error message.

## Related issue number (if any)
ENG-1359

## Checklist before requesting a review
- [x] I have created a descriptive PR title. The PR title should complete the sentence "This PR...".
- [x] I have performed a self-review of my code.
- [ ] I have included a small demo of the changes. For the UI, this would be a screenshot or a Loom video.
- [x] If this is a new feature, I have added unit tests and integration tests.
- [x] I have run the integration tests locally and they are passing.
- [x] I have run the linter script locally (See `python3 scripts/run_linters.py -h` for usage).
- [ ] All features on the UI continue to work correctly.
- [ ] Added one of the following CI labels:
    - `run_integration_test`: Runs integration tests
    - `skip_integration_test`: Skips integration tests (Should be used when changes are ONLY documentation/UI)

### CODE REVIEW

**Comment by vsreekanti** [comment]
This looks good to me. My only very slight concern is if the wording might make people think that all column types have to be strings? But I can't come up with anything clearer off the top of my head, so I'm happy with it as is.

**Comment by kenxu95** [comment]
I exactly copied the error that we get when trying to serialize a dataframe `to_parquet()`. I know we use json-serialization now, but there was some discussion about moving to a better serialization format like parquet in the future. So I figured if it's a constraint there, we might as well adopt it now and see if anyone complains.

File: src/python/aqueduct_executor/operators/function_executor/execute.py [hunk 1]
```diff
@@ -237,18 +228,19 @@ def run(spec: FunctionSpec) -> None:
                 # Cast np.bool_ to a bool.
                 result = bool(result)
             else:
-                handle_type_error_and_exit(spec, storage, exec_state, TIP_NOT_BOOL)
+                raise ExecFailureException(
```
Comment (likawind) [inline src/python/aqueduct_executor/operators/function_executor/execute.py:231]: I don't know if this is crucial, but it seems that we will lose all user printed logs for such failures?

**Review by likawind — APPROVED** [review]
Thanks for changing the pattern. My only question is to confirm with @vsreekanti that losing user logs under failures not caused by user function is acceptable, which I think so.

**Comment by vsreekanti** [comment]
It's not ideal, but it's fine for now.

File: src/python/aqueduct_executor/operators/function_executor/execute.py (hunk 1 above)
Comment (kenxu95) [reply src/python/aqueduct_executor/operators/function_executor/execute.py:231]: We won't lose any user logs. The except clause will add the user logs to the exec_state before writing. `exec_state.from_exception()` forces you to add user logs.

**Pull request merged**